### PR TITLE
Enable log_softmax and CrossEntropyLoss for bfloat16

### PR DESCRIPTION
**Pull request by hongzhen1**
Enabled torch.nn.functional.log_softmax and torch.nn.CrossEntropyLoss for bfloat16 data type.
In order to do that, following dependency have to be enabled.
- RNE (round to nearest even)
- AccumulateType
- bfloat16 arithmetic operator overload

Also, we implement std::numeric_limits fully support for bfloat16 data type

background for dependency:
- RNE vs truncate
From torch.nn.CrossEntropyLoss test. input_size=(128, 1000)
RNE result:
float    output:  tensor(7.3981, dtype=torch.float32, grad_fn=<NllLossBackward>)
bfloat16 output:  tensor(7.3125, dtype=torch.bfloat16, grad_fn=<NllLossBackward>)
truncate result:
float    output:  tensor(7.3981, dtype=torch.float32, grad_fn=<NllLossBackward>)
bfloat16 output:  tensor(5.8750, dtype=torch.bfloat16, grad_fn=<NllLossBackward>)

- scalar_t vs AccumulateType (AccumulateType of bfloat16 is float)
AccumulateType is essential to keep accuracy, especially for reduction related operation.
we have verified it with both local case and real topology. It turns out that bfloat16 type accumulator would cause huge relative error when elements number is large, even more than 50%.

### CODE REVIEW

**Comment by hongzhen1** [comment]
@gchanan @ezyang @jgong5 please have a review. Thanks

**Comment by hongzhen1** [comment]
@uyongw

**Comment by jgong5** [comment]
@izdeby @Jianhui-Li

**Comment by ezyang** [comment]
I think @izdeby is the correct reviewer for this PR

**Comment by hongzhen1** [comment]
@ezyang, Thanks for your reminding.
@izdeby, Please help review this, Thanks.

File: aten/src/ATen/nn_parse.py [hunk 1]
```diff
@@ -225,7 +225,7 @@ def unique_args(argslist):
     return result
 
 
-def function_info(name, arguments, cimpls, buffers, backends, inplace, scalar_check, backend_types):
+def function_info(name, arguments, cimpls, buffers, backends, inplace, scalar_check, backend_types, cpu_bf16):
```
Comment (izdeby) [inline aten/src/ATen/nn_parse.py]: Is there any better way than adding additional parameter specific to one type only? How is Half doing it?

File: c10/util/BFloat16.h [hunk 2]
```diff
@@ -43,6 +43,19 @@ namespace detail {
 
     return res >> 16;
   }
+
+  inline C10_HOST_DEVICE uint16_t round_to_nearest_even(float src) {
```
Comment (izdeby) [inline c10/util/BFloat16.h]: Can you, please, add some tests for this?

**Comment by izdeby** [comment]
Overall looks good to me except couple questions. 
Please, rebase and add some tests.

**Review by izdeby — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/SoftMax.cpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ void host_softmax(Tensor output, const Tensor& input, const int64_t dim) {
           for (int64_t d = 1; d < dim_size; d++)
             max_input = std::max(max_input, input_data[d * dim_stride]);
 
-          scalar_t tmpsum = 0;
+          acc_type<scalar_t, false> tmpsum = 0;
```
[the remark this answers is not in the document]
Comment (hongzhen1) [reply aten/src/ATen/native/SoftMax.cpp]: Accumulating operation (bf16 + bf16 + ...) w/ left to right order will introduce huge rounding error as the increase of  accumulating  number. we have done some experiments.  1000 data w/ Gaussian distribution will cause 50% relative error in real workload comparing with float accumulation.  you can  find similar operation in CUDA path for Half type.

File: c10/util/BFloat16.h (hunk 2 above)
Comment (hongzhen1) [reply c10/util/BFloat16.h]: RNE test is added into c10/test/util/bfloat16_test.cpp

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1280,12 +1280,14 @@
   named_guard: False
 
 - func: _log_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
+  cpu_bfloat16: True
```
[the remark this answers is not in the document]
Comment (hongzhen1) [reply aten/src/ATen/native/native_functions.yaml]: Removed

File: aten/src/ATen/nn_parse.py (hunk 1 above)
Comment (hongzhen1) [reply aten/src/ATen/nn_parse.py]: Hi, @izdeby, I have packed the parser interface with one better way. Please review again

**Review by izdeby — APPROVED** [review]

**Comment by gchanan** [comment]
What's the plan for bfloat16?  Why were these specific operators added?  Is it to support a certain model?

**Comment by Jianhui-Li** [comment]
@gchanan Intel’s upcoming Cooper Lake 14nm Intel Xeon® processor family will add Bfloat16 support, which provides 2x speedup for SIMD FMA instructions and 2x performance benefits on memory access. MKL-DNN v1.0 introduced bfloat16 support and expect more to come in the future releases. Comparing to FP32 baseline, we projected 1.6x+ performance benefit end to end for a wide range of vision models, and expect benefit for speech-recognition, recommendation engines and machine translation also.  Please refer to the RFC link https://github.com/pytorch/pytorch/issues/23509

**Comment by gchanan** [comment]
@Jianhui-Li thanks, but I don't think that answers "Why were these specific operators added? Is it to support a certain model?".

**Comment by jgong5** [comment]
>> Why were these specific operators added? Is it to support a certain model?

@gchanan Yes, we are going to support ResNet-50 training as the first step.

**Comment by hongzhen1** [comment]
@izdeby , thanks for your comments, and all changes have been approved by you. But Checking process suffers from remote machine timeout or network access issue. If you are convenient, could you help re-trigger these blocked checks and go through merge process ?

**Comment by hongzhen1** [comment]
> @pytorchbot retest this please

**Comment by hongzhen1** [comment]
Hi @izdeby, could we merge this PR firstly. Since this PR include some essential fundamental features, and it's valuable for enabling other BFloat16 ops. Thanks
